### PR TITLE
Fixed salt len

### PR DESCRIPTION
**Pull request by danielhuang**
The salt should be an array of length 16. That way, the length can be checked at compile time.

### CODE REVIEW

**Comment by MaximilianHollis** [comment]
Makes sense ! I had an extra allocation to pass a slice of the known salt

**Comment by YangJeffrey** [comment]
I agree 💯

File: src/lib.rs [hunk 1]
```diff
@@ -183,17 +181,23 @@ pub fn hash_with_result<P: AsRef<[u8]>>(password: P, cost: u32) -> BcryptResult<
...
 pub fn verify<P: AsRef<[u8]>>(password: P, hash: &str) -> BcryptResult<bool> {
+    use std::convert::TryInto;
+
     let parts = split_hash(hash)?;
     let salt = base64::decode_config(&parts.salt, base64::BCRYPT)?;
-    let generated = _hash_password(password.as_ref(), parts.cost, &salt)?;
+    let generated = _hash_password(
+        password.as_ref(),
+        parts.cost,
+        salt.try_into().expect("salt should be length 16"),
```
Comment (tarcieri) [inline src/lib.rs:199]: Won’t this panic if the salt is the wrong length?

Comment (danielhuang) [reply src/lib.rs:199]: Good catch! I'll be publishing a fix.

Comment (danielhuang) [reply src/lib.rs:199]: The reason I used `expect` was due to an assumption that a panic would have occurred if a slice of length different from 16 would be passed through the previous implementation.

**Comment by danielhuang** [comment]
I added a new variant to `BcryptError`. This is probably not the best way—please let me know.

**Comment by Keats** [comment]
Looks like a legit test failure

**Comment by danielhuang** [comment]
Tests are passing now. `TryInto` was imported from `std`, instead of `core`.

**Review by Keats — APPROVED** [review]

File: src/bcrypt.rs [hunk 2]
```diff
@@ -13,12 +13,12 @@ fn setup(cost: u32, salt: &[u8], key: &[u8]) -> Blowfish {
     state
 }
 
-pub fn bcrypt(cost: u32, salt: &[u8], password: &[u8], output: &mut [u8]) {
-    assert!(salt.len() == 16);
+pub fn bcrypt(cost: u32, salt: [u8; 16], password: &[u8]) -> [u8; 24] {
```
Comment (Numenorean) [inline src/bcrypt.rs:16]: consider referencing salt instead giving ownership? because further in code we are still taking a reference

**Pull request merged**